### PR TITLE
Branch detection updates

### PR DESCRIPTION
**Pull request by JelmerBot**
This PR solves #660, adding `labels` and `probability` parameters to `BranchDetector.fit()` that override the input HDBSCAN object. Cases where overridden clusters form multiple connected components in the minimum spanning tree are detected. The component labels are returned as branch labels in that case. The condensed and linkage trees for those clusters are set to None, allowing scripts to detect what happened.

While working on this PR, I noticed that the branching code could be simplified extensively. This PR revert some of the changes I made when I introduced the branch detection code. I also found and fixed some issues with the hierarchy simplification code that applies a persistence threshold and added a persistence threshold parameter to the clustering code. 

Finally, I made small changes in `_hdbscan_boruvka.pyx` to expose the computed core distances and neighbours. This allows me to use the implementation in another project.

### CODE REVIEW

File: hdbscan/branch_data.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import numpy as np
+from sklearn.neighbors import KDTree, BallTree
+from .dist_metrics import DistanceMetric
+
+
+class BranchDetectionData(object):
+    """Input data for branch detection functionality.
+
+    Recreates and caches internal data structures from the clustering stage.
+
```
Comment (JelmerBot) [inline hdbscan/branch_data.py:10]: Moved `BranchDetectionData` to a new file so `branches.py` can import `hdbscan_.py` without cyclical imports.

File: hdbscan/hdbscan_.py [hunk 2]
```diff
@@ -1215,12 +1236,12 @@ def fit(self, X, y=None):
...
-                self._finite_index = get_finite_row_indices(X)
-                clean_data = X[self._finite_index]
+                finite_index = get_finite_row_indices(X)
+                clean_data = X[finite_index]
                 internal_to_raw = {
-                    x: y for x, y in zip(range(len(self._finite_index)), self._finite_index)
+                    x: y for x, y in zip(range(len(finite_index)), finite_index)
                 }
-                outliers = list(set(range(X.shape[0])) - set(self._finite_index))
+                outliers = list(set(range(X.shape[0])) - set(finite_index))
```
Comment (JelmerBot) [inline hdbscan/hdbscan_.py:1244]: I now recover the finite index from the condensed tree, so it does not have to be stored explicitly anymore.

File: hdbscan/tests/test_hdbscan.py [hunk 3]
```diff
@@ -9,8 +9,7 @@
 from sklearn.utils.estimator_checks import check_estimator
 from sklearn.utils._testing import (
     assert_array_equal,
-    assert_array_almost_equal,
-    assert_raises,
+    assert_array_almost_equal
```
Comment (JelmerBot) [inline hdbscan/tests/test_hdbscan.py:12]: `assert_raises` gives import error on CI/CD. I replaced it with `pytest.raises` in all tests.

**Comment by lmcinnes** [comment]
This looks great. Let me know when you are ready to have it merged.

**Review by lmcinnes — APPROVED** [review]

**Comment by lmcinnes** [comment]
Thanks for all your work on this!
